README.md
# firebase_ui

[![Build Status](https://travis-ci.org/fidelisa/flutter_plugins.svg?branch=master)](https://travis-ci.org/fidelisa/flutter_plugins)
[![pub package](https://img.shields.io/pub/v/firebase_ui.svg)](https://pub.dartlang.org/packages/firebase_ui)

Fork of https://pub.dartlang.org/packages/flutter_firebase_ui

Old package wasn't being maintained so I created this one to update.

firebase_ui is a library that helps you to quickly implement firebase authentication.
It provides UI for the common identity providers like Facebook, Google, Twitter and email.

Note: this plugin use firebase_auth you must configure it as describe [here](https://pub.dartlang.org/packages/firebase_auth).


![screenshot](https://i.gyazo.com/cbb58dcfcbb16c45d314e8e885e84c6c.png)

![screenshot](https://i.gyazo.com/32b79e6ecc14d98558f25688cdcf5b9e.png)

Show [firebase ui](https://github.com/Maliffic/firebase_ui/tree/master/example) example.

## Getting Started

For help getting started with Flutter, view our online [documentation](http://flutter.io/).

For help on editing package code, view the [documentation](https://flutter.io/developing-packages/).


CHANGELOG.md
## [1.0.10] - 31/12/2019
* allow guest login

## [1.0.9] - 30/12/2019
* update dependencies

## [1.0.8] - 02/10/2019
* update dependencies

## [1.0.7] - 21/5/2019
* fixed padding

## [1.0.6] - 21/5/2019
* added better padding options
* swapped listView for a column
* changed twitter cancel error message

## [1.0.5] - 21/5/2019
* update dependencies
* fixing twitter


## [1.0.1] - 29/4/2019
* update dependencies
* added custon padding to buttoms
* added avoidBottomInset

## [0.0.19] - 19/12/2018
* update dependencies
* added show/hide app barq

## [0.0.13] - 19/12/2018
* update dependencies
* add twitter

## [0.0.12] - 27/09/2018
* after trouble sign in pop

## [0.0.11] - 27/09/2018
* add function signOutProviders

## [0.0.10] - 26/09/2018
* fixed empty null on Providers
* allow for usage with landscape orientation

## [0.0.9] - 07/09/2018
* add mail verification option

## [0.0.8] - 17/07/2018
* improve UI

## [0.0.7] - 04/07/2018
* fixe sign in with mail

## [0.0.6] - 30/05/2018
* improve show dialog error

## [0.0.5] - 29/05/2018
* show error occured on unknown errors
* display email if no displayName (example)
* secure providers test, firebase_auth behavior change
* pop sign up !

## [0.0.4] - 09/05/2018
* Upgrade to Google Services 3.2.1 and Google Sign In 3.0.3

## [0.0.3] - 23/03/2018
* dart 2 fix

## [0.0.2] - 20/03/2018

* improve ui
* change firebase_auth version

## [0.0.1] - 19/03/2018

* initial release


lib/utils.dart
import 'dart:async';

import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_ui/l10n/localization.dart';
import 'package:flutter/material.dart';
import 'package:flutter_facebook_login/flutter_facebook_login.dart';
import 'package:google_sign_in/google_sign_in.dart';

enum ProvidersTypes { email, google, facebook, twitter, guest, phone }

final GoogleSignIn googleSignIn = new GoogleSignIn();
final FacebookLogin facebookLogin = new FacebookLogin();

ProvidersTypes stringToProvidersType(String value) {
  if (value.toLowerCase().contains('facebook')) return ProvidersTypes.facebook;
  if (value.toLowerCase().contains('google')) return ProvidersTypes.google;
  if (value.toLowerCase().contains('password')) return ProvidersTypes.email;
  if (value.toLowerCase().contains('twitter')) return ProvidersTypes.twitter;
  if (value.toLowerCase().contains('guest')) return ProvidersTypes.guest;
//TODO  if (value.toLowerCase().contains('phone')) return ProvidersTypes.phone;
  return null;
}

// Description button
class ButtonDescription extends StatelessWidget {
  final String label;
  final Color labelColor;
  final Color color;
  final String logo;
  final IconData icon;
  final String name;
  final VoidCallback onSelected;

  const ButtonDescription(
      {@required this.label,
      @required this.name,
      this.logo,
      this.icon,
      this.onSelected,
      this.labelColor = Colors.grey,
      this.color = Colors.white});

  ButtonDescription copyWith({
    String label,
    Color labelColor,
    Color color,
    String logo,
    IconData icon,
    String name,
    VoidCallback onSelected,
  }) {
    return new ButtonDescription(
        label: label ?? this.label,
        labelColor: labelColor ?? this.labelColor,
        color: color ?? this.color,
        logo: logo ?? this.logo,
        icon: icon ?? this.icon,
        name: name ?? this.name,
        onSelected: onSelected ?? this.onSelected);
  }

  @override
  Widget build(BuildContext context) {
    VoidCallback _onSelected = onSelected ?? () => {};
    return new RaisedButton(
        color: color,
        child: new Row(
          children: <Widget>[
            new Container(
              padding: const EdgeInsets.fromLTRB(16.0, 16.0, 32.0, 16.0),
              child: icon != null ? Icon(icon, color: labelColor, size: 30 ) : Image.asset('assets/$logo', package: 'firebase_ui'),
            ),
            new Expanded(
              child: new Text(
                label,
                style: new TextStyle(color: labelColor),
              ),
            ),
          ],
        ),
        onPressed: _onSelected);
  }
}

Map<ProvidersTypes, ButtonDescription> providersDefinitions(BuildContext context) => {
      ProvidersTypes.facebook: new ButtonDescription(
          color: const Color.fromRGBO(59, 87, 157, 1.0),
          logo: "fb-logo.png",
          label: FFULocalizations.of(context).signInFacebook,
          name: "Facebook",
          labelColor: Colors.white),
      ProvidersTypes.google: new ButtonDescription(
          color: Colors.white,
          logo: "go-logo.png",
          label: FFULocalizations.of(context).signInGoogle,
          name: "Google",
          labelColor: Colors.black54),
      ProvidersTypes.email: new ButtonDescription(
          color: const Color.fromRGBO(219, 68, 55, 1.0),
          logo: "email-logo.png",
          label: FFULocalizations.of(context).signInEmail,
          name: "Email",
          labelColor: Colors.white),
      ProvidersTypes.twitter: new ButtonDescription(
          color: const Color.fromRGBO(29, 161, 242, 1.0),
          logo: "twitter-logo.png",
          label: FFULocalizations.of(context).signInTwitter,
          name: "Twitter",
          labelColor: Colors.white),
      ProvidersTypes.guest: new ButtonDescription(
          color: const Color.fromRGBO(244, 180, 0, 1.0),
          icon: Icons.person,
          label: FFULocalizations.of(context).signInGuest,
          name: "guest",
          labelColor: Colors.white),
    };

Future<Null> showErrorDialog(BuildContext context, String message, {String title}) {
  return showDialog<Null>(
    context: context,
    barrierDismissible: false, // user must tap button!
    builder: (BuildContext context) => new AlertDialog(
      title: title != null ? new Text(title) : null,
      content: new SingleChildScrollView(
        child: new ListBody(
          children: <Widget>[
            new Text(message ?? FFULocalizations.of(context).errorOccurred),
          ],
        ),
      ),
      actions: <Widget>[
        new FlatButton(
          child: new Row(
            children: <Widget>[
              new Text(FFULocalizations.of(context).cancelButtonLabel),
            ],
          ),
          onPressed: () {
            Navigator.of(context).pop();
          },
        ),
      ],
    ),
  );
}

Future<void> signOutProviders() async {
  var currentUser = await FirebaseAuth.instance.currentUser();
  if (currentUser != null) {
    await signOut(currentUser.providerData);
  }

  return await FirebaseAuth.instance.signOut();
}

Future<dynamic> signOut(Iterable providers) async {
  return Future.forEach(providers, (p) async {
    switch (p.providerId) {
      case 'facebook.com':
        await facebookLogin.logOut();
        break;
      case 'google.com':
        await googleSignIn.signOut();
        break;
    }
  });
}
